src/OrbitQt/HistogramWidget.cpp

```diff
@@ -127,61 +121,175 @@ static void DrawHistogram(QPainter& painter, const QPoint& axes_intersection,
 
-  DrawTitle(function_name_.value(), painter);
-}
\ No newline at end of file
+  if (IsSelectionActive()) {
+    DrawTitle(painter, absl::StrFormat("Selection over %u points", histogram.data_set_size));
+  } else {
+    DrawTitle(painter, function_name_.value());
+  }
+
+  if (selection_start_pixel && selection_current_pixel) {
+    DrawSelection(painter, *selection_start_pixel, *selection_current_pixel, axes_intersection,
+                  vertical_axis_length);
+  }
+}
+
+void HistogramWidget::mousePressEvent(QMouseEvent* event) {
+  if (histogram_stack_.empty()) return;
+
+  const int pixel_x = event->x();
+  selection_start_pixel = pixel_x;
+  selection_current_pixel = pixel_x;
+
+  update();
+}
+
+[[nodiscard]] static uint64_t LocationToValue(int pos_x, int width, int margin, uint64_t min_value,
+                                              uint64_t max_value) {
+  if (pos_x <= margin) return 0;
+  if (pos_x > width - margin) return max_value + 1;
+
+  const int location = pos_x - margin;
+  const int histogram_width = width - 2 * margin;
+  const uint64_t value_range = max_value - min_value;
+  return min_value +
+         static_cast<uint64_t>(static_cast<double>(location) / histogram_width * value_range);
+}
+
+void HistogramWidget::mouseReleaseEvent(QMouseEvent* /* event*/) {
+  if (histogram_stack_.empty()) return;
+
+  if (selection_start_pixel && selection_current_pixel) {
+    // if it wasn't a drag, but just a click, go one level of selections up
+    if (*selection_start_pixel == *selection_current_pixel) {
+      if (IsSelectionActive()) histogram_stack_.pop();
+      selection_start_pixel.reset();
+      selection_current_pixel.reset();
+      update();
+      return;
+    }
+
+    uint64_t min =
+        LocationToValue(*selection_start_pixel, Width(), WidthMargin(), MinValue(), MaxValue());
+    uint64_t max =
+        LocationToValue(*selection_current_pixel, Width(), WidthMargin(), MinValue(), MaxValue());
+    if (min > max) {
+      std::swap(min, max);
+    }
+
+    const auto min_it = std::lower_bound(data_->begin(), data_->end(), min);
+    const auto max_it = std::upper_bound(data_->begin(), data_->end(), max);
+    const auto selection = absl::Span<const uint64_t>(&*min_it, std::distance(min_it, max_it));
+
+    auto histogram = orbit_statistics::BuildHistogram(selection);
+    if (histogram) {
+      histogram_stack_.push(std::move(*histogram));
+    }
+
+    selection_start_pixel.reset();
+    selection_current_pixel.reset();
+  }
+
+  update();
+}
+
+void HistogramWidget::mouseMoveEvent(QMouseEvent* event) {
+  if (!selection_start_pixel) return;
+
+  selection_current_pixel = event->x();
+
+  update();
+}
+
+bool HistogramWidget::IsSelectionActive() const { return histogram_stack_.size() > 1; }
+
+uint64_t HistogramWidget::MinValue() const {
+  return IsSelectionActive() ? histogram_stack_.top().min : 0;
+}
+
+uint64_t HistogramWidget::MaxValue() const { return histogram_stack_.top().max; }
```

Comment:
Not really a big deal since they are private, but why are they asymmetric? In other words: Could they be implemented symmetrically?

Reply:
Because I've chosen to plot the histograms in asymmetric manner. The initial histogram always starts with zero.